core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/AvroUpsertMonitor.scala

```diff
@@ -194,63 +219,116 @@ class AvroUpsertMonitorActor(
+            case Failure(t) => publishMessageToUpdateImportStatus(attributes.importId, Option(status), ImportStatuses.Error, Option(t.getMessage))
+          }
+        }
+        case Some(status) => publishMessageToUpdateImportStatus(attributes.importId, Option(status), ImportStatuses.Error, Option("Import was not ready for upsert"))
+        case None => publishMessageToUpdateImportStatus(attributes.importId, None, ImportStatuses.Error, Option("Import status not found"))
+      }
+    } yield ()
+  }
+
+
+  private def markImportAsDone(importId: UUID, workspaceName: WorkspaceName, userInfo: UserInfo) = {
+    for {
+      importStatus <- importServiceDAO.getImportStatus(importId, workspaceName, userInfo)
+      _ <- importStatus match {
+        case Some(status) if status == ImportStatuses.Upserting => publishMessageToUpdateImportStatus(importId, Option(status), ImportStatuses.Done, None)
+        case Some(status) => publishMessageToUpdateImportStatus(importId, Option(status), ImportStatuses.Error, Option("Upsert finished, but import status was not correct."))
+        case None => publishMessageToUpdateImportStatus(importId, None, ImportStatuses.Error, Option("Import status not found"))
+      }
+    } yield ()
+  }
+
+  private def publishMessageToUpdateImportStatus(importId: UUID, currentImportStatus: Option[ImportStatus], newImportStatus: ImportStatus, errorMessage: Option[String]) = {
+    val updateImportStatus = UpdateImportStatus(importId.toString, newImportStatus.toString, currentImportStatus.map(_.toString), errorMessage)
+    val attributes = updateImportStatus.toJson.asJsObject.fields.map { case (attName, attValue) =>
+      // JsValue -> String, adds extra quotes for some reason
+      (attName, attValue.compactPrint.replace("\"","")) }
+    pubSubDao.publishMessages(importStatusPubSubTopic, Seq(GooglePubSubDAO.MessageRequest("", attributes)))
+  }
+
+
+  private def initUpsert(upsertFile: String, jobId: UUID, ackId: String, workspaceName: WorkspaceName, userInfo: UserInfo): Future[Unit] = {
+    for {
+      //ack the response after we load the json into memory. pro: don't have to worry about ack timeouts for long operations, con: if someone restarts rawls here the upsert is lost
+      upsertJson <- readUpsertObject(upsertFile)
       ackResponse <- acknowledgeMessage(ackId)
-      upsertResults <-
-        avroUpsertJson.grouped(batchSize).toList.traverse { upsertBatch =>
+      ws = workspaceService.apply(userInfo)
+      upsertResults <- {
+        upsertJson.grouped(batchSize).toList.traverse { upsertBatch =>
           logger.info(s"starting upsert for $jobId with ${upsertBatch.size} entities ...")
-          IO.fromFuture(IO(workspaceService.apply(petUserInfo).batchUpdateEntities(WorkspaceName(avroMetadataJson.namespace, avroMetadataJson.name), upsertBatch, true)))
-        }.unsafeToFuture
+          IO.fromFuture(IO(ws.batchUpdateEntities(workspaceName, upsertBatch, true)))
+        }.unsafeToFuture}
```

Comment:
for consideration ... it would be nice to know how many batches are in this job. It's extra runtime code; is it worth it for the log  message?
```
        val batches = upsertJson.grouped(batchSize).zipWithIndex.toList
        val numBatches = batches.size
        batches.traverse {
          case (upsertBatch, idx) =>
            logger.info(s"starting upsert $idx of $batchSize for $jobId with ${upsertBatch.size} entities ...")
            IO.fromFuture(IO(ws.batchUpdateEntities(workspaceName, upsertBatch, true)))
        }.unsafeToFuture
```

Reply:
This would log for every single entity update - so if I uploaded a pfb with 5000 entities I would log 5000 lines. That would make the logs pretty ugly...
If you want to know the batch sizes - we could log that instead of every single thing in the batch

Reply:
Actually, this is fine, nvm, deleted my previous comment :)